README.md
# reflex-sdl2-test
Test programs and examples using reflex-sdl2


CHANGELOG.md
# Revision history for reflex-sdl2-test

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


pong/Pong.hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE DeriveFunctor #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecursiveDo #-}
{-# LANGUAGE ScopedTypeVariables #-}
module Main where

import Control.Applicative
import Control.Monad
import Data.Functor

import Control.Lens ((^.), over)
import Reflex
import Reflex.SDL2
import System.Random (randomRIO)

import Reflex.SDL2.Layers
import Reflex.SDL2.Keys

gameSize :: Num a => V2 a
gameSize = V2 800 600

main :: IO ()
main = do
  initializeAll
  let windowCfg = defaultWindow{ windowInitialSize = gameSize }
  window <- createWindow "PONG - WS / arrow keys" windowCfg
  r <- createRenderer window (-1) defaultRenderer
  rendererDrawBlendMode r $= BlendAlphaBlend
  hostLayers r app
  destroyRenderer r
  destroyWindow window
  quit

app :: (ReflexSDL2 t m, MonadLayer t m, MonadSample t (Performable m)) => m ()
app = do
  dPaddleL <- verticalPaddle defaultPaddle{paddleXOffset = 100, paddleYRange = 500}
                =<< axisKeyControl ScancodeW  ScancodeS
  dPaddleR <- verticalPaddle defaultPaddle{paddleXOffset = 700, paddleYRange = 500}
                =<< axisKeyControl ScancodeUp ScancodeDown
  (dBall, ePoint) <- ball dPaddleL dPaddleR 20
  dScore <- trackScores ePoint
  eQuit <- getQuitEvent
  performEvent_ $ eQuit $> do
    Scores scoreL scoreR <- sample (current dScore)
    liftIO . putStrLn $ concat
      [ "Final score: Left "
      , show scoreL
      , " - "
      , show scoreR
      , " Right"
      ]
  shutdownOn =<< delay 0 =<< getQuitEvent

data PaddleConfig = PaddleConfig
  { paddleSpeed   :: Double
  , paddleHeight  :: Int
  , paddleXOffset :: Int
  , paddleYRange  :: Int
  } deriving (Eq, Show)

defaultPaddle :: PaddleConfig
defaultPaddle = PaddleConfig
  { paddleSpeed = 300
  , paddleHeight = 100
  , paddleXOffset = 0
  , paddleYRange = 0
  }

clamp :: Ord a => a -> a -> a -> a
clamp lo hi x
  | x < lo = lo
  | x > hi = hi
  | otherwise = x

verticalPaddle :: (ReflexSDL2 t m, MonadLayer t m)
               => PaddleConfig
               -> Dynamic t Sign
               -> m (Dynamic t (Rectangle Int))
verticalPaddle cfg dControl = do
  let startY = maxY / 2
      minY = 0
      maxY = fromIntegral (paddleYRange cfg)
      dSpeed = signedValue (paddleSpeed cfg) <$> dControl
      paddleSize = V2 20 (paddleHeight cfg)
  eTick <- getDeltaTickEvent
  let eFracTick = fmap ((/1000) . fromIntegral) eTick
      eUpdate = attachWith (*) (current dSpeed) eFracTick
  dPaddleY <- fmap round <$> foldDyn (\dy y -> clamp minY maxY $ dy + y) startY eUpdate
  drawLayer $ ffor dPaddleY \y r -> do
    rendererDrawColor r $= V4 255 255 0 255
    fillRect r (Just (fromIntegral <$>
                        Rectangle (P (V2 (paddleXOffset cfg - 10) y))
                                          paddleSize))
  pure $ ffor dPaddleY \y -> Rectangle (P (V2 (paddleXOffset cfg - 10) y)) paddleSize

data Player = LeftPlayer | RightPlayer

data Scores = Scores !Int !Int

trackScores :: ReflexSDL2 t m => Event t Player -> m (Dynamic t Scores)
trackScores = foldDyn addPoint (Scores 0 0)
  where
    addPoint LeftPlayer (Scores l r) = Scores (l+1) r
    addPoint RightPlayer (Scores l r) = Scores l (r+1)

data Edge = LeftEdge | RightEdge | TopEdge | BottomEdge

edgePlayer :: Edge -> Maybe Player
edgePlayer LeftEdge  = Just LeftPlayer
edgePlayer RightEdge = Just RightPlayer
edgePlayer _         = Nothing

ball :: forall t m. (ReflexSDL2 t m, MonadLayer t m)
     => Dynamic t (Rectangle Int)
     -> Dynamic t (Rectangle Int)
     -> Int
     -> m (Dynamic t (V2 Int), Event t Player)
ball dLeftPaddle dRightPaddle rad = do
  let radius = fromIntegral rad
      colliding leftPaddle rightPaddle pos
        = LeftEdge <$ guard (rectsIntersect (fromIntegral <$> leftPaddle) (ballRect pos))
          <|> RightEdge <$ guard (rectsIntersect (fromIntegral <$> rightPaddle) (ballRect pos))
      dPaddleCollider = zipDynWith colliding dLeftPaddle dRightPaddle

      startPos = gameSize / 2

      ballSizeHalf = V2 radius radius
      ballSize = 2 * ballSizeHalf

      ballRect pos = Rectangle (P (pos - ballSizeHalf)) ballSize
      
      reflect LeftEdge   = over _x abs
      reflect RightEdge  = over _x (negate . abs)
      reflect TopEdge    = over _y abs
      reflect BottomEdge = over _y (negate . abs)

      randomVel :: MonadIO m' => m' (V2 Double)
      randomVel = liftIO do
        u <- angle <$> randomRIO (0, 2*pi)
        l <- randomRIO (70, 130)
        pure (l *^ u)
  
  eFracTick  <- fmap ((/1000) . fromIntegral) <$> getDeltaTickEvent 
  ePostBuild <- getPostBuild

  let dIsCollidingPaddle dBallPos = zipDynWith ($) dPaddleCollider dBallPos
      ePaddleCollision   dBallPos = fforMaybe (updated (dIsCollidingPaddle dBallPos)) id
      eVBoundsCollision  dBallPos = fforMaybe (updated (dBallPos))
                                      \pos -> let closeTop = pos ^._y < radius
                                                  closeBot = pos ^._y > gameSize ^._y - radius
                                              in TopEdge <$ guard closeTop
                                                <|> BottomEdge <$ guard closeBot
      eHBoundsCollision  dBallPos = fforMaybe (updated dBallPos)
                                      \pos -> let closeL = pos ^._x < 0
                                                  closeR = pos ^._x > gameSize ^._x
                                              in RightPlayer <$ guard closeL
                                                <|> LeftPlayer <$ guard closeR
      
  rec dBallPos :: Dynamic t (V2 Double) <- foldDyn ($) startPos =<< delay 0 ePosUpdate
      
      let ePosUpdate :: Event t (V2 Double -> V2 Double)
          ePosUpdate = leftmost [eStart, eReset, eIntegral]

          eIntegral = (+) <$> attachWith (^*) (current dBallVel) eFracTick
          eStart    = const startPos <$ ePostBuild
          eReset    = const startPos <$ eHBoundsCollision dBallPos
      
      dBallVel :: Dynamic t (V2 Double) <- foldDyn ($) 0 =<< delay 0 eVelUpdate
      
      eVelUpdate :: Event t (V2 Double -> V2 Double) <- do
        let eBounce = mergeWith (.)
              [ reflect <$> ePaddleCollision dBallPos
              , reflect <$> eVBoundsCollision dBallPos
              ]
        eStart <- performEvent (randomVel <$ ePostBuild)
        eReset <- performEvent (randomVel <$ eHBoundsCollision dBallPos)
        pure $ leftmost [const <$> eStart, const <$> eReset, eBounce]
  drawLayer $ ffor dBallPos \pos r -> do
    rendererDrawColor r $= V4 255 0 0 255
    fillRect r (Just (round <$> ballRect pos))
  pure ( fmap round <$> dBallPos
       , eHBoundsCollision dBallPos)

rectsIntersect :: (Num a, Ord a) => Rectangle a -> Rectangle a -> Bool
rectsIntersect (Rectangle (P p) (V2 x y)) r2 = any (`inRect` r2)
  [ p
  , p + V2 x 0
  , p + V2 x y
  , p + V2 0 y
  ]

inRect :: (Num a, Ord a) => V2 a -> Rectangle a -> Bool
inRect q (Rectangle (P p) (V2 dx dy)) = inX && inY
  where
    V2 x y = q - p
    inX = 0 <= x && x <= dx
    inY = 0 <= y && y <= dy


moving-block/MovingBlock.hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE ViewPatterns #-}
module Main where

import Control.Applicative
import Control.Monad
import Data.Fixed (mod')
import Data.Foldable
import Data.Monoid
import Foreign.C.Types
import System.Environment

import CoercibleUtils
import Reflex
import Reflex.SDL2
import Time.Repeatedly

import Reflex.SDL2.Keys
import Reflex.SDL2.Layers

gameSize :: Num a => V2 a
gameSize = V2 800 600

main :: IO ()
main = do
  initializeAll
  let windowCfg = defaultWindow{ windowInitialSize = gameSize }
  window <- createWindow "a moving block - use arrow keys" windowCfg
  r <- createRenderer window (-1) defaultRenderer
  rendererDrawBlendMode r $= BlendAlphaBlend
  hostLayers r app
  destroyRenderer r
  destroyWindow window
  quit


app :: (ReflexSDL2 t m, MonadLayer t m) => m ()
app = do
  _ <- movableBox arrowKeys 50 100
  shutdownOn =<< delay 0 =<< getQuitEvent

data MoveKeys = MoveKeys
  { mkUp, mkDown, mkLeft, mkRight :: Scancode
  }

arrowKeys :: MoveKeys
arrowKeys = MoveKeys
  { mkUp    = ScancodeUp
  , mkDown  = ScancodeDown
  , mkLeft  = ScancodeLeft
  , mkRight = ScancodeRight
  }

getMoveDirection :: (ReflexSDL2 t m) => MoveKeys -> m (Dynamic t (V2 Sign))
getMoveDirection MoveKeys{..} = do
  dVertical <- axisKeyControl mkUp mkDown
  dHorizontal <- axisKeyControl mkLeft mkRight
  pure $ zipDynWith V2 dHorizontal dVertical

wrapV2 :: Real a => V2 a -> V2 a -> V2 a
wrapV2 = liftA2 (flip mod')

movableBox :: (ReflexSDL2 t m, MonadLayer t m) => MoveKeys -> Double -> V2 Int -> m (Dynamic t (V2 Int))
movableBox keys speed halfSize = do
  dMoveDir <- getMoveDirection keys
  eTick <- getDeltaTickEvent
  ePB <- getPostBuild
  let eUpdate = attachWith update (current dMoveDir) $ leftmost [eTick, 0 <$ ePB]
      update dir dt = \pos -> wrapV2 gameSize (pos + fmap (signedValue speed) dir * fromIntegral dt / 1000)
  dBoxPos <- (fmap . fmap) round <$> foldDyn id 0 eUpdate
  drawLayer $ ffor dBoxPos \pos r -> do
    rendererDrawColor r $= V4 255 0 0 255
    fillRect r (Just (fromIntegral <$> Rectangle (P (pos - halfSize)) (halfSize*2)))
  pure dBoxPos

getPeriodicTick :: (ReflexSDL2 t m) => Double -> m (Event t ())
getPeriodicTick freq = do
  let timer cb = void . liftIO $ asyncRepeatedly (toRational freq) (cb ())
  eBuilt <- getPostBuild
  performEventAsync (timer <$ eBuilt)
  


reflextra/Reflex/SDL2/Keys.hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE FlexibleContexts #-}
module Reflex.SDL2.Keys where

import Control.Monad

import Reflex
import Reflex.SDL2

getIsKeyDown :: ReflexSDL2 t m => Scancode -> m (Dynamic t Bool)
getIsKeyDown key = do
  eKeyAll <- getKeyboardEvent
  let eMyKey = fforMaybe eKeyAll \e -> do
        guard $ key == keysymScancode (keyboardEventKeysym e)
             && not (keyboardEventRepeat e)
        pure (Pressed == keyboardEventKeyMotion e)
  holdDyn False eMyKey

data Sign = Negative | Zero | Positive
  deriving (Eq, Ord, Show)

instance Semigroup Sign where
  Zero <> x = x
  x <> Zero = x
  x <> y
    | x == y    = x
    | otherwise = Zero

instance Monoid Sign where
  mempty = Zero

signedValue :: Num a => a -> Sign -> a
signedValue a Negative = -a
signedValue _ Zero     = 0
signedValue a Positive = a

invertSign :: Sign -> Sign
invertSign Negative = Positive
invertSign Zero     = Zero
invertSign Positive = Positive

axisKeyControl :: ReflexSDL2 t m => Scancode -> Scancode -> m (Dynamic t Sign)
axisKeyControl kNeg kPos = do
  dNeg <- getIsKeyDown kNeg
  dPos <- getIsKeyDown kPos
  let combine True  False = Negative
      combine False True  = Positive
      combine _     _     = Zero
  pure $ zipDynWith combine dNeg dPos


moving-block/Main.hs
{-# LANGUAGE BlockArguments #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE ViewPatterns #-}
module Main where

import Control.Applicative
import Control.Monad
import Data.Fixed (mod')
import Data.Foldable
import Data.Monoid
import Foreign.C.Types
import System.Environment

import CoercibleUtils
import Reflex
import Reflex.SDL2
import Time.Repeatedly

gameSize :: Num a => V2 a
gameSize = V2 800 600

main :: IO ()
main = do
  freq <- ffor getArgs \case
    [f] -> read f :: Double
    _   -> 20
  initializeAll
  let windowCfg = defaultWindow{ windowInitialSize = gameSize }
  window <- createWindow "reflex-sdl2-test" windowCfg
  r <- createRenderer window (-1) defaultRenderer
  rendererDrawBlendMode r $= BlendAlphaBlend
  hostLayers r app
  destroyRenderer r
  destroyWindow window
  quit

type Layer m = Renderer -> Performable m ()
type MonadLayer t m = MonadDynamicWriter t [Layer m] m

drawLayer :: (ReflexSDL2 t m, MonadLayer t m) => Dynamic t (Layer m) -> m ()
drawLayer = tellDyn . fmap pure

hostLayers :: Renderer -> DynamicWriterT Spider [Layer ConcreteReflexSDL2] ConcreteReflexSDL2 () -> IO ()
hostLayers r guest = host do
  (_, dLayers) <- runDynamicWriterT guest
  ePB <- getPostBuild
  let eDisplay = leftmost [updated dLayers, [] <$ ePB]
  performEvent_ $ ffor eDisplay \layers -> do
    rendererDrawColor r $= V4 0 0 0 255
    clear r
    traverse_ ($ r) layers
    present r

app :: (ReflexSDL2 t m, MonadLayer t m) => m ()
app = do
  _ <- movableBox arrowKeys 50 100
  shutdownOn =<< delay 0 =<< getQuitEvent

getIsKeyDown :: ReflexSDL2 t m => Scancode -> m (Dynamic t Bool)
getIsKeyDown key = do
  allKeyEvts <- getKeyboardEvent
  holdDyn False $ fforMaybe allKeyEvts \e -> do
    guard $ key == keysymScancode (keyboardEventKeysym e)
         && not (keyboardEventRepeat e)
    pure (Released == keyboardEventKeyMotion e)

data MoveKeys = MoveKeys
  { mkUp, mkDown, mkLeft, mkRight :: Scancode
  }

arrowKeys :: MoveKeys
arrowKeys = MoveKeys
  { mkUp    = ScancodeUp
  , mkDown  = ScancodeDown
  , mkLeft  = ScancodeLeft
  , mkRight = ScancodeRight
  }

getMoveDirection :: (ReflexSDL2 t m, Num a) => MoveKeys -> m (Dynamic t (V2 a))
getMoveDirection MoveKeys{..} = do
  dUp    <- fmap (whenNum (V2   0   1 )) <$> getIsKeyDown mkUp
  dDown  <- fmap (whenNum (V2   0 (-1))) <$> getIsKeyDown mkDown
  dLeft  <- fmap (whenNum (V2   1   0 )) <$> getIsKeyDown mkLeft
  dRight <- fmap (whenNum (V2  (-1) 0 )) <$> getIsKeyDown mkRight
  pure $ ala (Sum . Ap) foldMap [dUp, dDown, dLeft, dRight]

whenNum :: Num a => a -> Bool -> a
whenNum a = \b -> if b then a else 0

wrapV2 :: Real a => V2 a -> V2 a -> V2 a
wrapV2 = liftA2 (flip mod')

movableBox :: (ReflexSDL2 t m, MonadLayer t m) => MoveKeys -> Double -> V2 Int -> m (Dynamic t (V2 Int))
movableBox keys speed halfSize = do
  dMoveDir <- getMoveDirection keys
  eTick <- getDeltaTickEvent
  ePB <- getPostBuild
  let eUpdate = attachWith update (current dMoveDir) $ leftmost [eTick, 0 <$ ePB]
      update dir dt = \pos -> wrapV2 gameSize (pos + dir * V2 speed speed * fromIntegral dt / 1000)
  dBoxPos <- (fmap . fmap) round <$> foldDyn id 0 eUpdate
  drawLayer $ ffor dBoxPos \pos r -> do
    rendererDrawColor r $= V4 255 0 0 255
    fillRect r (Just (fromIntegral <$> Rectangle (P (pos - halfSize)) (halfSize*2)))
  pure dBoxPos

getPeriodicTick :: (ReflexSDL2 t m) => Double -> m (Event t ())
getPeriodicTick freq = do
  let timer cb = void . liftIO $ asyncRepeatedly (toRational freq) (cb ())
  eBuilt <- getPostBuild
  performEventAsync (timer <$ eBuilt)
  